File: backend/poker/engine/player.py
from typing import Optional

from .action import Action


class Player:
    name: str
    stack: int
    seat_number: Optional[int]
    folded: bool
    last_action: Optional[Action]
    sitting_out: bool

    def __init__(self, _name: str) -> None:
        self.name = _name
        self.stack = 0
        self.seat_number = None
        self.folded = False
        self.last_action = None
        self.sitting_out = False

    def make_bet(self, bet_size: int) -> int:
        assert self.stack >= bet_size, "Player may not bet amount larger than their stack"
        self.stack -= bet_size
        return bet_size

    def receive_pot(self, pot_size: int) -> int:
        self.stack += pot_size
        return pot_size

    def __repr__(self):
        return str(self.__dict__)

File: engine/player.go
package engine

import (
	"errors"
	"github.com/simster7/PocketRockets/api/v1"
)

type Player struct {
	Name       string
	Stack      int
	SeatNumber int
	Folded     bool
	IsAllIn    bool
	LastAction Action
	SittingOut bool
}

func NewPlayer(name string, stack int) Player {
	return Player{
		Name:       name,
		Stack:      stack,
		SeatNumber: -1,
		Folded:     false,
		LastAction: Action{},
		SittingOut: false,
	}
}

func (p *Player) MakeBet(betSize int) error {
	if p.Stack < betSize {
		return errors.New("player may not bet amount larger than their stack")
	}
	p.Stack -= betSize
	return nil
}

func (p *Player) ReceivePot(potSize int) {
	p.Stack += potSize
}

func (p *Player) SetLastAction(lastAction Action) {
	p.LastAction = lastAction
}

func (p *Player) SetFolded(folded bool) {
	p.Folded = folded
}

func (p *Player) SetIsAllIn(isAllIn bool) {
	p.IsAllIn = isAllIn
}

func (p *Player) GetMessage() *v1.Player {
	return &v1.Player{
		Name:       p.Name,
		Stack:      int32(p.Stack),
		SeatNumber: int32(p.SeatNumber),
		Folded:     p.Folded,
		IsAllIn:    p.IsAllIn,
		SittingOut: p.SittingOut,
		LastAction: p.LastAction.GetMessage(),
	}
}
